internal/observability/metrics.go

```diff
@@ -0,0 +1,57 @@
+package observability
+
+import (
+	"context"
+	"sync"
+)
+
+type contextKey string
+
+const keyMetricsRecorder contextKey = "metricsRecorder"
+
+type (
+	Metric interface {
+		IsMetric()
+	}
+
+	MetricsRecorder interface {
+		Metrics() []Metric
+		RecordMetric(m Metric)
+	}
+)
+
+type recorder struct {
+	m  []Metric
+	mu sync.Mutex
+}
+
+func newRecorder() MetricsRecorder {
+	return &recorder{
+		m: make([]Metric, 0),
+	}
+}
+
+func RecorderFromContext(ctx context.Context) MetricsRecorder {
+	r := ctx.Value(keyMetricsRecorder)
+	mr, ok := r.(MetricsRecorder)
+	if !ok {
+		panic("developer error")
```

Comment:
actually, I think this should return a dummy `MetricsRecorder` if the lookup fails:
```go
type noopRecorder struct{}

func (noopRecorder) RecordMetric(m Metric) {}

func RecorderFromContext(ctx context.Context) MetricsRecorder {
    if mr, ok := ctx.Value(keyMetricsRecorder).(MetricsRecorder); ok {
        return mr
    }
    return noopRecorder{}
}
```

Reply:
I considered this when writing it but decided it would be safer to throw a `panic` for the time being as a `noopRecorder` might hide cases where we simply forgot to add one. Added the `noopRecorder` though 👍